components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/CallBackValidator.java

```diff
@@ -0,0 +1,112 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.endpoint.client;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.IdentityProviderProperty;
+import org.wso2.carbon.identity.mgt.endpoint.IdentityManagementEndpointConstants;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Validates the call back URL passed in the recovery request.
+ */
+public class CallBackValidator {
+
+    private static final Log log = LogFactory.getLog(CallBackValidator.class);
+
+    /**
+     * This method is to validate the callback URL in the request with the configured one.
+     *
+     * @param callbackURL  passed in the request
+     * @param tenantDomain of the user
+     * @return the status of the validation
+     * @throws IdentityRecoveryException
+     */
+    public boolean isValidCallbackURL(String callbackURL, String tenantDomain, boolean isUserPortalURL)
+            throws IdentityRecoveryException {
+
+        if (isUserPortalURL) {
+            if (log.isDebugEnabled()) {
+                log.debug("Callback URL is equal to the user portal URL: " + isUserPortalURL);
+            }
+            return true;
+        }
+
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+            if (log.isDebugEnabled()) {
+                log.debug("Tenant domain is considered as super tenant domain: " + tenantDomain);
+            }
+        }
+
+        IdentityProvider residentIdP;
+        try {
+            residentIdP = IdentityProviderManager.getInstance().getResidentIdP(tenantDomain);
+        } catch (IdentityProviderManagementException e) {
+            throw new IdentityRecoveryException("Error occurred while reading the resident IdP for the tenant domain : "
+                    + tenantDomain, e);
+        }
+
+        IdentityProviderProperty[] idpProperties = null;
+        if (residentIdP != null) {
+            idpProperties = residentIdP.getIdpProperties();
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Resident identity provider is not found for the tenant domain: " + tenantDomain);
+            }
+        }
+
+        String callbackRegex = null;
+        if (idpProperties != null) {
+            for (IdentityProviderProperty property : idpProperties) {
+                if (property.getName().equals(IdentityManagementEndpointConstants.UserInfoRecovery
```

Comment:
shall we switch the two arguments of equals statement since RECOVERY_CALLBACK_REGEX is a constant here? This will avoid possible NPEs.

Reply:
fixed